### ISSUE TITLE
Error aws->postgresql unexpected EOF

### ISSUE THREAD

**Issue by castaples** [cloudquery/cloudquery#6431]
After upgrading to the latest version CQ and AWS/Postgres plugins I get the following output in the logs after the sync exits after only 7 mins running (usual runs for close to 3 hours) having gathered credentials and started syncing tables.

```
Error: failed to write for aws->postgresql: failed to CloseAndRecv client: rpc error: code = Internal desc = Context done: context canceled and failed to wait for plugin: failed to execute batch: unexpected EOF
```

We've gone from CQ `1.3.7` to `2.0.31`, AWS from `v3.5.0` to `v9.1.1` and Postgres from `v1.3.9` to `v2.0.1`

We're running CQ in a docker container on AWS ECS with Aurora Serverless v1 Postgres (pre-scaled to 16 ACU before sync called), this has worked fairly well for some time on a daily schedule. We have 350+ AWS accounts with only a few tables in the CQ config ignore list, mainly 'aws_inspector2_findings' and all the lightsail tables.

Please let me know what else you'd need to diagnose this.

**Comment by hermanschaaf** [comment]
Hi @castaples, thanks for reaching out! Other than the output of the CLI, could you tell us if there are any other error-level messages in the `cloudquery.log` file?

**Comment by castaples** [comment]
Yes so there are a lot, if I filter out the 403 errors (we probably need to update the role CQ is using to give it a few more permissions) which we had previously and it was working, I can see a lot relating to CodePipeline, for example:
```json
{
    "level": "error",
    "module": "aws-src",
    "client": "************:eu-west-1",
    "error": "operation error CodePipeline: ListWebhooks, https response error StatusCode: 0, RequestID: , canceled, context canceled",
    "message": "table resolver finished with error",
    "table": "aws_codepipeline_webhooks",
    "time": "2023-01-06T09:22:43Z"
}
```

**Comment by yevgenypats** [comment]
Hi @castaples , thx for the report! Can you please try postgresql `v2.0.3` - https://github.com/cloudquery/cloudquery/releases/tag/plugins-destination-postgresql-v2.0.3 and let us know if that worked or still the same behaviour ?

**Comment by castaples** [comment]
I've tried again with CQ `2.0.31`, AWS `v9.1.1` and Postgres `v2.0.3`, unfortunately I get the same problem, unexpected EOF error.

```json
{
    "level": "error",
    "module": "cli",
    "error": "failed to write for aws->postgresql: failed to CloseAndRecv client: rpc error: code = Internal desc = Context done: context canceled and failed to wait for plugin: failed to execute batch: unexpected EOF",
    "time": "2023-01-09T08:48:51Z",
    "message": "exiting with error"
}
```

I've been though the logs and tried to filter out the noise to find anything that might be useful, these stood out to me as they had some detail in the error messages.

```json
{
    "level": "error",
    "module": "aws-src",
    "client": "************:ap-southeast-3",
    "error": "operation error ApiGatewayV2: GetApis, https response error StatusCode: 400, RequestID: 32284d50-cad3-4b5a-84f5-3f94ff856287, deserialization failed, failed to decode response body, invalid character '\\'' in string escape code",
    "message": "table resolver finished with error",
    "table": "aws_apigatewayv2_apis",
    "time": "2023-01-09T08:44:10Z"
}
```

```json
{
    "level": "error",
    "module": "aws-src",
    "client": "************:ca-central-1",
    "error": "operation error ApiGatewayV2: GetDomainNames, https response error StatusCode: 200, RequestID: cbb34ac1-9660-42bf-9aae-21563cd470cb, deserialization failed, failed to decode response body with invalid JSON, expected json.Number to be of type string, got string instead",
    "message": "table resolver finished with error",
    "table": "aws_apigatewayv2_domain_names",
    "time": "2023-01-09T08:44:14Z"
}
```

```json
{
    "level": "error",
    "module": "aws-src",
    "client": "************:us-east-1",
    "error": "operation error Backup: GetBackupVaultNotifications, https response error StatusCode: 400, RequestID: 9bbd25a9-dfc7-4806-92a8-b12b20f8eae0, api error ERROR_2106: Failed reading notifications from database for Backup vault Default",
    "message": "column resolver finished with error",
    "table": "aws_backup_vaults",
    "time": "2023-01-09T08:45:42Z"
}
```

This final entry occurs just before the Postgres unexpected EOF error

```json
{
    "level": "info",
    "module": "cli",
    "address": "/tmp/cq-VgkcXiwkhVWaTpsA.sock",
    "message": "Got interrupt. Source plugin server shutting down",
    "time": "2023-01-09T08:48:51Z"
}
```

Please let me know if you want me to try anything else or look through the logs for something specific, I'll endeavour to get the CQ role permissions updated to rule out lack of permissions causing this issue.

**Comment by erezrokah** [comment]
Hi @castaples, thanks for digging through the logs and providing these extra errors, super valuable.
The error you're seeing is coming from either https://github.com/cloudquery/cloudquery/blob/d766e56fb8519c378b0ba10cef734032796f9d4d/plugins/destination/postgresql/client/write.go#L78 or https://github.com/cloudquery/cloudquery/blob/d766e56fb8519c378b0ba10cef734032796f9d4d/plugins/destination/postgresql/client/write.go#L93

The errors you've discovered (except the last one) are related to the source plugin. However we're looking for the destination plugin detailed error (if present).
The last error is expected as we stop the sync if there's an error write a batch to Postgres. Looks like that in this case we are unable to extract useful information from the received error.

A possible way to get more about the failure is to use another destination and see if that works or we get more information (for the sake of debugging).
For example see the SQLite destination docs [here](https://www.cloudquery.io/docs/plugins/destinations/sqlite/overview)

**Comment by castaples** [comment]
I've tried several times to get SQLite v1.1.5 (CQ 2.1 & AWS v10.1.0) working but keep getting the following error:

```json
{
    "level": "error",
    "module": "cli",
    "error": "failed to create destination plugin client for sqlite: failed to start destination plugin .cq/plugins/destination/cloudquery/sqlite/v1.1.5/plugin: fork/exec .cq/plugins/destination/cloudquery/sqlite/v1.1.5/plugin: no such file or directory",
    "time": "2023-01-11T14:42:59Z",
    "message": "exiting with error"
}
```
Config
```yml
kind: source
spec:
  ## Required. name of the plugin to use.
  ## This should be unique if you have number of aws plugins.
  name: 'aws'
  ## Optional. Where to search for the plugin. Default: "github". Options: "github", "local", "grpc"
  # registry: "github"

  ## Path for the plugin.
  ## If registry is "github" path should be "repo/name"
  ## If registry is "local", path is path to binary. If "grpc" then it should be address of the plugin (usually useful in debug).
  path: 'cloudquery/aws'

  ## Required. Must be a specific version starting with v, e.g. v1.2.3
  ## checkout latest versions here https://github.com/cloudquery/cloudquery/releases?q=plugins-source-aws&expanded=true
  version: ${CQ_VAR_CLOUD_QUERY_AWS_PLUGIN_VERSION}

  ## Optional. Default: ["*"] - all tables. We recommend to specify specific tables that you need to sync as this
  ## will reduce the amount of data synced and improve performance.
  ## See all tables: https://github.com/cloudquery/cloudquery/blob/main/plugins/source/aws/docs/tables/README.md
  # tables: ["*"]

  skip_tables:
    [
      'aws_inspector2_findings',
      'aws_lightsail_alarms',
      'aws_lightsail_buckets',
      'aws_lightsail_bucket_access_keys',
      'aws_lightsail_certificates',
      'aws_lightsail_container_services',
      'aws_lightsail_container_service_deployments',
      'aws_lightsail_container_service_images',
      'aws_lightsail_databases',
      'aws_lightsail_database_parameters',
      'aws_lightsail_database_events',
      'aws_lightsail_database_log_events',
      'aws_lightsail_database_snapshots',
      'aws_lightsail_disks',
      'aws_lightsail_disk_snapshot',
      'aws_lightsail_distributions',
      'aws_lightsail_instances',
      'aws_lightsail_instance_port_states',
      'aws_lightsail_instance_snapshots',
      'aws_lightsail_load_balancers',
      'aws_lightsail_load_balancer_tls_certificates',
      'aws_lightsail_static_ips',
    ]

  ## Required. all destinations you want to sync data to.
  destinations: ['sqlite']

  # resource_concurrency: 100000

  spec:
    ## Optional. plugin specific configuration
    ## By default will use the current aws credentials available (just like AWS CLI)
    ## See all available options here: https://github.com/cloudquery/cloudquery/blob/main/plugins/source/aws/docs/configuration.md
    org:
      member_role_name: ${CQ_VAR_AWS_CLOUD_QUERY_MEMBER_ROLE_NAME}
      admin_account:
        id: 'UK_Consolidated'
        role_arn: ${CQ_VAR_AWS_CLOUD_QUERY_ADMIN_ROLE_ARN}
      member_trusted_principal:
        id: 'CloudQueryTaskRole'
        role_arn: ${CQ_VAR_AWS_CLOUD_QUERY_ROLE_ARN}
---
kind: destination
spec:
  name: sqlite
  path: cloudquery/sqlite
  version: 'v1.1.5'
  spec:
    connection_string: /app/db.sql
```

As the versions of CQ and AWS plugin have changed since I last tried Postgres, I will try again with the latest versions.

**Comment by hermanschaaf** [comment]
@castaples Are there any other errors before the `no such file or directory` one? It seems like it's failing to download or save the sqlite plugin into the container for some reason 🤔

**Comment by castaples** [comment]
@hermanschaaf full log output below:

```json
{
    "level": "info",
    "module": "cli",
    "args": [
        "/config/config.yml"
    ],
    "time": "2023-01-11T14:42:56Z",
    "message": "Loading spec(s)"
}
{
    "level": "info",
    "module": "cli",
    "address": "/tmp/cq-wGCjPFZIEamCvPBP.sock",
    "message": "Source plugin server listening",
    "time": "2023-01-11T14:42:58Z"
}
{
    "level": "info",
    "module": "cli",
    "grpc.code": "OK",
    "grpc.component": "server",
    "grpc.method": "GetProtocolVersion",
    "grpc.method_type": "unary",
    "grpc.service": "proto.Source",
    "grpc.start_time": "2023-01-11T14:42:59Z",
    "grpc.time_ms": "0.03",
    "message": "started call",
    "peer.address": "@",
    "protocol": "grpc",
    "time": "2023-01-11T14:42:59Z"
}
{
    "level": "info",
    "module": "cli",
    "grpc.code": "OK",
    "grpc.component": "server",
    "grpc.method": "GetProtocolVersion",
    "grpc.method_type": "unary",
    "grpc.service": "proto.Source",
    "grpc.start_time": "2023-01-11T14:42:59Z",
    "grpc.time_ms": "0.055",
    "message": "finished call",
    "peer.address": "@",
    "protocol": "grpc",
    "time": "2023-01-11T14:42:59Z"
}
{
    "level": "info",
    "module": "cli",
    "source": "aws",
    "destinations": [
        "sqlite"
    ],
    "sync_time": "2023-01-11T14:42:59Z",
    "time": "2023-01-11T14:42:59Z",
    "message": "Start sync"
}
{
    "level": "info",
    "module": "cli",
    "source": "aws",
    "destinations": [
        "sqlite"
    ],
    "sync_time": "2023-01-11T14:42:59Z",
    "time": "2023-01-11T14:42:59Z",
    "message": "End sync"
}
{
    "level": "info",
    "module": "cli",
    "address": "/tmp/cq-wGCjPFZIEamCvPBP.sock",
    "message": "Got interrupt. Source plugin server shutting down",
    "time": "2023-01-11T14:42:59Z"
}
Error: failed to create destination plugin client for sqlite: failed to start destination plugin .cq/plugins/destination/cloudquery/sqlite/v1.1.5/plugin: fork/exec .cq/plugins/destination/cloudquery/sqlite/v1.1.5/plugin: no such file or directory
{
    "level": "error",
    "module": "cli",
    "error": "failed to create destination plugin client for sqlite: failed to start destination plugin .cq/plugins/destination/cloudquery/sqlite/v1.1.5/plugin: fork/exec .cq/plugins/destination/cloudquery/sqlite/v1.1.5/plugin: no such file or directory",
    "time": "2023-01-11T14:42:59Z",
    "message": "exiting with error"
}
```

**Comment by hermanschaaf** [comment]
That is quite strange. Let's set aside the SQLite issue for now, since you're actually trying to use PostgreSQL. What resources are allocated to the container? Previous reports of the "Unexpected EOF" error have usually been that the environment ran out of memory, so I wonder if that could be the case here as well.

**Comment by castaples** [comment]
I have run into issues with CPU/memory usage with older versions, so we're currently running a fairly large container with 16 vCPU and 120 GB of memory in ECS Fargate, RDS Aurora Serverless v1 Postgres 11.13 is pre-warmed to 16 ACU.

Looking at container insights for the last time it failed I can see that at the point of the "Unexpected EOF" the CPU usage was only 15.5% and memory 7.5%, this appears to be fairly typical for the other times it failed in the last week of trying newer versions.

Typically with CQ `1.3.7`, AWS `v3.5.0` and Postgres `v1.3.9` (our last known good configuration) we'd see a huge spike in CPU and memory usage at the start, for example it'd idle for the first 5ish minutes, spike to 98% CPU and 70% memory for 10 minutes, then CPU would almost idle and memory slowly decline for the next 3 hours of running until it completes. This influenced the size of container we've used to ensure we don't get near 100% memory usage, as that would cause the container to stop.

**Comment by hermanschaaf** [comment]
@castaples I'm looking into this... I'll try and see what changed between these versions, and I notice there have been a few [Unexpected EOF errors](https://github.com/jackc/pgx/issues?q=is%3Aissue+unexpected+EOF+) reported for the pgx library that we use, so there might be something there. My current theory is that it's related to connection management with Aurora Serverless, which may be different from normal Postgres installations.

**Comment by hermanschaaf** [comment]
@castaples While I continue investigating connection management, could you also try running a sync with the new versions, but skipping all the new AWS tables that have been added since `v3.5.0`? This should give us a more apples-to-apples comparison. I auto-generated a list for you here by comparing the list of tables for the two versions:

```
- aws_account_alternate_contacts
- aws_account_contacts
- aws_amp_rule_groups_namespaces
- aws_amp_workspaces
- aws_apigatewayv2_api_integration_responses
- aws_apigatewayv2_api_route_responses
- aws_apprunner_auto_scaling_configurations
- aws_apprunner_connections
- aws_apprunner_custom_domains
- aws_apprunner_observability_configurations
- aws_apprunner_operations
- aws_apprunner_vpc_connectors
- aws_apprunner_vpc_ingress_connections
- aws_appstream_app_blocks
- aws_appstream_application_fleet_associations
- aws_appstream_applications
- aws_appstream_directory_configs
- aws_appstream_fleets
- aws_appstream_image_builders
- aws_appstream_images
- aws_appstream_stack_entitlements
- aws_appstream_stack_user_associations
- aws_appstream_stacks
- aws_appstream_usage_report_subscriptions
- aws_appstream_users
- aws_athena_data_catalog_database_tables
- aws_cloudwatchlogs_resource_policies
- aws_config_config_rule_compliances
- aws_config_config_rules
- aws_docdb_certificates
- aws_docdb_cluster_parameter_groups
- aws_docdb_cluster_parameters
- aws_docdb_cluster_snapshots
- aws_docdb_clusters
- aws_docdb_engine_versions
- aws_docdb_event_categories
- aws_docdb_event_subscriptions
- aws_docdb_events
- aws_docdb_global_clusters
- aws_docdb_instances
- aws_docdb_orderable_db_instance_options
- aws_docdb_pending_maintenance_actions
- aws_docdb_subnet_groups
- aws_ecr_repository_image_scan_findings
- aws_elasticsearch_packages
- aws_elasticsearch_versions
- aws_elasticsearch_vpc_endpoints
- aws_elastictranscoder_pipeline_jobs
- aws_elastictranscoder_pipelines
- aws_elastictranscoder_presets
- aws_elbv2_listener_certificates
- aws_eventbridge_api_destinations
- aws_eventbridge_archives
- aws_eventbridge_connections
- aws_eventbridge_endpoints
- aws_eventbridge_event_sources
- aws_eventbridge_replays
- aws_frauddetector_batch_imports
- aws_frauddetector_batch_predictions
- aws_frauddetector_detectors
- aws_frauddetector_entity_types
- aws_frauddetector_event_types
- aws_frauddetector_external_models
- aws_frauddetector_labels
- aws_frauddetector_model_versions
- aws_frauddetector_models
- aws_frauddetector_outcomes
- aws_frauddetector_rules
- aws_frauddetector_variables
- aws_fsx_file_caches
- aws_glue_database_table_indexes
- aws_glue_registry_schema_versions
- aws_iam_ssh_public_keys
- aws_identitystore_group_memberships
- aws_identitystore_groups
- aws_identitystore_users
- aws_kafka_cluster_operations
- aws_kafka_clusters
- aws_kafka_configurations
- aws_kafka_nodes
- aws_kms_key_grants
- aws_kms_key_policies
- aws_lambda_layer_version_policies
- aws_mq_broker_configuration_revisions
- aws_mwaa_environments
- aws_neptune_global_clusters
- aws_organization_resource_policies
- aws_organizations
- aws_quicksight_analyses
- aws_quicksight_dashboards
- aws_quicksight_data_sets
- aws_quicksight_data_sources
- aws_quicksight_folders
- aws_quicksight_group_members
- aws_quicksight_groups
- aws_quicksight_ingestions
- aws_quicksight_templates
- aws_quicksight_users
- aws_ram_principals
- aws_ram_resource_share_associations
- aws_ram_resource_share_invitations
- aws_ram_resource_share_permissions
- aws_ram_resource_shares
- aws_ram_resource_types
- aws_ram_resources
- aws_rds_cluster_parameters
- aws_rds_engine_versions
- aws_redshift_cluster_parameters
- aws_savingsplans_plans
- aws_scheduler_schedule_groups
- aws_scheduler_schedules
- aws_servicecatalog_portfolios
- aws_servicecatalog_products
- aws_servicecatalog_provisioned_products
- aws_servicequotas_quotas
- aws_servicequotas_services
- aws_ses_active_receipt_rule_sets
- aws_ses_configuration_set_event_destinations
- aws_ses_configuration_sets
- aws_ses_contact_lists
- aws_ses_custom_verification_email_templates
- aws_ses_identities
- aws_ssm_associations
- aws_ssm_compliance_summary_items
- aws_ssm_instance_patches
- aws_ssm_inventories
- aws_ssm_inventory_schemas
- aws_ssm_patch_baselines
- aws_ssoadmin_account_assignments
- aws_ssoadmin_instances
- aws_ssoadmin_permission_sets
- aws_stepfunctions_state_machines
- aws_timestream_databases
- aws_timestream_tables
- aws_xray_resource_policies
```

By the way, in newer versions you should also be able to use wildcard matching, so you could skip all of lightsail with `aws_lightsail*`

**Comment by hermanschaaf** [comment]
Thanks for trying and providing the details @castaples ! Too bad it's not good news, looks like the issue really isn't down to any specific resources we added. And sorry yeah, I see my list of tables included two that were only added a few hours ago and not officially released yet (I was working against the main branch) :) I'll continue digging into potential connection management issues

**Comment by hermanschaaf** [comment]
@castaples I still suspect it's due to a change in connection handling (either in our code or between versions in pgx), but it's hard to prove without the ability to reproduce this issue myself. I've looked through the code but nothing jumps out at me. 

If am right, however, I believe your sync should succeed when using an older version of the postgresql plugin (e.g. `v1.3.9` like you were before) while keeping `aws` and the CLI at more recent versions. Could you test this theory for me and report back what you find? If that doesn't work, you can also try `postgres v1.8.0`, the version right before we switched to pgx v5. Knowing whether this still worked would be really valuable.

**Comment by castaples** [comment]
@hermanschaaf I've run with CQ `2.1.0`, AWS `v10.1.0` and Postgres `v1.3.9` its show table migration completes, then sync starts but now immediately gets a "EOF" error (before it would start gathering data from accounts before failing)

```json
{
    "level": "info",
    "module": "cli",
    "grpc.code": "OK",
    "grpc.method": "Migrate",
    "grpc.method_type": "unary",
    "grpc.service": "proto.Destination",
    "grpc.start_time": "2023-01-13T10:40:37Z",
    "grpc.time_ms": "2179.342",
    "kind": "server",
    "message": "finished server unary call",
    "system": "grpc",
    "time": "2023-01-13T10:40:39Z"
}
{
    "level": "info",
    "module": "cli",
    "source": "aws",
    "destinations": [
        "postgresql"
    ],
    "num_tables": 317,
    "time_took": 2.182915301,
    "time": "2023-01-13T10:40:39Z",
    "message": "End migration"
}
{
    "level": "info",
    "module": "cli",
    "source": "aws",
    "destinations": [
        "postgresql"
    ],
    "time": "2023-01-13T10:40:39Z",
    "message": "Start fetching resources"
}
{
    "level": "info",
    "module": "cli",
    "grpc.code": "OK",
    "grpc.component": "server",
    "grpc.method": "Sync2",
    "grpc.method_type": "server_stream",
    "grpc.service": "proto.Source",
    "grpc.start_time": "2023-01-13T10:40:39Z",
    "grpc.time_ms": "0.058",
    "message": "started call",
    "peer.address": "@",
    "protocol": "grpc",
    "time": "2023-01-13T10:40:39Z"
}
{
    "level": "info",
    "module": "cli",
    "address": "/tmp/cq-bcItdNTUGlkAyZCk.sock",
    "message": "Got interrupt. Destination plugin server shutting down",
    "time": "2023-01-13T10:40:39Z"
}
{
    "level": "info",
    "module": "cli",
    "source": "aws",
    "destinations": [
        "postgresql"
    ],
    "sync_time": "2023-01-13T10:40:35Z",
    "time": "2023-01-13T10:40:39Z",
    "message": "End sync"
}
{
    "level": "info",
    "module": "cli",
    "address": "/tmp/cq-cchwDHFmiAPYMguN.sock",
    "message": "Got interrupt. Source plugin server shutting down",
    "time": "2023-01-13T10:40:39Z"
}
Error: failed to write for aws->postgresql: failed to send tables: EOF
{
    "level": "error",
    "module": "cli",
    "error": "failed to write for aws->postgresql: failed to send tables: EOF",
    "time": "2023-01-13T10:40:39Z",
    "message": "exiting with error"
}
```

I've then tried CQ `2.1.0`, AWS `v10.1.0` and Postgres `v1.8.0` which got further but still runs into the "Unexpected EOF" (I've just filtered those logs with level = "error")

```json
{
    "level": "error",
    "module": "cli",
    "error": "failed to execute batch: unexpected EOF",
    "message": "failed to wait",
    "time": "2023-01-13T10:54:39Z"
}
{
    "level": "error",
    "module": "cli",
    "grpc.code": "Unknown",
    "grpc.component": "server",
    "grpc.error": "context canceled",
    "grpc.method": "Write2",
    "grpc.method_type": "client_stream",
    "grpc.service": "proto.Destination",
    "grpc.start_time": "2023-01-13T10:51:08Z",
    "grpc.time_ms": "210258.52",
    "message": "finished call",
    "peer.address": "@",
    "protocol": "grpc",
    "time": "2023-01-13T10:54:39Z"
}
{
    "level": "error",
    "module": "aws-src",
    "client": "************:ap-south-1",
    "error": "operation error AccessAnalyzer: ListAnalyzers, https response error StatusCode: 0, RequestID: , canceled, context canceled",
    "message": "table resolver finished with error",
    "table": "aws_accessanalyzer_analyzers",
    "time": "2023-01-13T10:54:39Z"
}
Error: failed to write for aws->postgresql: failed to CloseAndRecv client: rpc error: code = Canceled desc = context canceled
{
    "level": "error",
    "module": "cli",
    "error": "failed to write for aws->postgresql: failed to CloseAndRecv client: rpc error: code = Canceled desc = context canceled",
    "time": "2023-01-13T10:54:39Z",
    "message": "exiting with error"
}
```

Sorry its taken so long to get these to you, please let me know if you would like me to try anything else.

**Comment by hermanschaaf** [comment]
Thanks again @castaples, really appreciate it. What's most interesting to me here is that the versions of pgx we use didn't change between postgresql plugin versions 1.3.9 and 1.8.0 (both used v4.17.2). So that seems to rule out that it could have been a change in our upstream dependency.

If you don't mind me asking one more thing: if you revert to all the previously known versions (CQ 1.3.7, AWS v3.5.0 and Postgres v1.3.9), does the sync still complete successfully? I would just like to rule out that it's something that changed in the environment. If we can rule that out, then we have a fairly small set of code changes to examine carefully and figure out why the unexpected EOF started happening.

**Comment by castaples** [comment]
@hermanschaaf Yes reverting to the previous versions means the sync then completes successfully.

**Comment by hermanschaaf** [comment]
So, I've been digging into this some more...

I tried to reproduce the issue by creating an Aurora Serverless instance (v2, AWS didn't allow me to create V1) and connecting to it from EC2. I'm afraid I wasn't able to reproduce the issue. The sync ran to completion in 11 minutes in my case, though it is testing against a much smaller cloud footprint than yours. I'm using the same plugin versions from when the issue was created (AWS v9.1.1, postgresql v2.0.1) with the most recent CLI.

I also tried setting the source `concurrency` to a lower value so that the sync would take longer, hoping to see if this might result in a connection being interrupted, but even after 20 minutes it was still running.

I then went through the log messages carefully again, and it seems like the `Unexpected EOF` is happening at a point where the destination plugin server is already shutting down (`failed to CloseAndRecv client`). I believe this can only happen if either 1) all resources have been fetched and the channel closed, or 2) the gRPC connection got interrupted for some reason, most probably because the destination plugin process was shut down. TL;DR I think the `Unexpected EOF` error is a red herring.

@castaples Would you perhaps be up for doing a Zoom debugging session with me and @yevgenypats some time? If so, you can set it up here https://calendly.com/yevgenyp

**Comment by castaples** [comment]
@hermanschaaf meeting scheduled, please let me know if you want me to prepare anything in advance.

**Comment by hermanschaaf** [comment]
@castaples Great, thanks! I think we'd mostly like to see your setup and how you run the CLI, where logs go, metrics for the container, what the configuration looks like, etc. So if you have an environment where you can reproduce the issue that you can show us, that would be amazing. Then hopefully we can spot something, or at least suggest some new things to try :)

**Comment by castaples** [comment]
@yevgenypats and @hermanschaaf thank you both for your time earlier in the week, I've tried running the sync with concurrency set at 250000 and 100000 but still get the same error as before and still no indication in container insights of high memory usage before the error. I will look to setup an EC2 instance to run from to allow for better monitoring of what's going on at the point of failure, however it may be some time before I can get to this.

Is there anything else you can suggest I try with the current setup (ECS/Aurora Serverless) to save time? I did have a look through the docs after our call and noticed there is a generic reference/setting for destination plugins for batch size https://www.cloudquery.io/docs/reference/destination-spec#batch_size would this have any bearing on the Postgres plugin if set?

**Comment by yevgenypats** [comment]
Hi @castaples , thanks for reporting the experiment. Yes definitely this will have effect and can have affect on the postgres side of things. The default as far as I can see in the code (we need to add it to the docs) is `10000` so maybe you can do `2000` be sure to put the `batch_size` in the upper `spec` for `destination` not in the `postgresql` `spec`.

Once you try on ec2, and if it will behave the same way we will add some retries on the Postgres plugin side but it will be much easier to debug it that way.

**Comment by castaples** [comment]
@hermanschaaf Apologies for the length of time its taken but I have now resolved the issue. I did initially setup an ec2 instance with local Postgres DB with as close to the same cpu/memory as the ECS container, same Postgres version and latest CQ/plugin versions, which I successfully ran a sync across all of our accounts. I then tried again with our original setup, ECS Fargate with Aurora Serverless v1 Postgres 11.16, this failed as before. I then truncated the DB and tried again, this time the sync worked, however subsequent syncs failed intermittently. I have now migrated to Aurora Serverless v2 Postgres 14.6, with latest CQ/plugins sync is now successfully running each day for a week now. I have come to the conclusion it was either the DB config/infrastructure or (less likely) the data, causing the error. Sorry I can't really offer much more insight but I'll close this issue now, thanks again for your support
